test/left_recursion/left_recursion_test.go

```diff
@@ -0,0 +1,197 @@
+			if err != nil {
+				t.Fatalf(
+					"for input %q got error: %s, but expect to parse without errors",
+					testCase.expr, err)
+			}
+			expr, ok := res.(string)
+			if !ok {
+				t.FailNow()
+			}
+			if expr != testCase.want.expr {
+				t.Fatalf(
+					"for input %q\ngot result: %q,\nbut expect: %q",
+					testCase.expr, expr, testCase.want.expr)
+			}
+		})
+
+		t.Run(testCase.name+" optimezed", func(t *testing.T) {
+			t.Parallel()
+
+			resLR, err := optimizedleftrecursion.Parse("", []byte(testCase.expr))
+			if err != nil {
+				t.Fatalf(
+					"for input %q got error: %s, but expect to parse without errors",
+					testCase.expr, err)
+			}
+			exprLR, ok := resLR.(string)
+			if !ok {
+				t.FailNow()
+			}
+			if exprLR != testCase.want.expr {
+				t.Fatalf(
+					"for input %q\ngot result: %q,\nbut expect: %q",
+					testCase.expr, exprLR, testCase.want.expr)
+			}
+			res, err := optimizedwithoutleftrecursion.Parse("", []byte(testCase.expr))
+			if err != nil {
+				t.Fatalf(
+					"for input %q got error: %s, but expect to parse without errors",
+					testCase.expr, err)
+			}
+			expr, ok := res.(string)
+			if !ok {
+				t.FailNow()
+			}
+			if expr != testCase.want.expr {
+				t.Fatalf(
+					"for input %q\ngot result: %q,\nbut expect: %q",
+					testCase.expr, expr, testCase.want.expr)
+			}
+		})
+	}
+}
+
+func FuzzLeftRecursionParse(f *testing.F) {
+	chars := []byte("0123456789+-/*%")
+
+	f.Fuzz(func(t *testing.T, bytes []byte) {
+		data := make([]byte, 0, len(bytes))
+		for _, b := range bytes {
+			data = append(data, chars[int(b)%len(chars)])
+		}
+		resLR, errLR := leftrecursion.Parse("", data)
+		res, err := withoutleftrecursion.Parse("", data)
+		if err != nil || errLR != nil {
+			if err == nil || errLR == nil {
+				t.Fatalf(
+					"for input %q\ngot error: %q,\nbut expect: %q",
+					data, errLR, err)
+			}
+			return
+		}
+		exprLR, okLR := resLR.(string)
+		if !okLR {
+			t.FailNow()
+		}
+		expr, ok := res.(string)
+		if !ok {
+			t.FailNow()
+		}
+		if expr != exprLR {
+			t.Fatalf(
+				"for input %q\ngot result: %q,\nbut expect: %q",
+				data, exprLR, expr)
+		}
+	})
+}
+
+func FuzzLeftRecursionParseOptimized(f *testing.F) {
+	chars := []byte("0123456789+-/*%")
+
+	f.Fuzz(func(t *testing.T, bytes []byte) {
+		data := make([]byte, 0, len(bytes))
+		for _, b := range bytes {
+			data = append(data, chars[int(b)%len(chars)])
+		}
+		resLR, errLR := optimizedleftrecursion.Parse("", data)
+		res, err := optimizedwithoutleftrecursion.Parse("", data)
+		if err != nil || errLR != nil {
+			if err == nil || errLR == nil {
+				t.Fatalf(
+					"for input %q\ngot error: %q,\nbut expect: %q",
+					data, errLR, err)
+			}
+			return
+		}
+		exprLR, okLR := resLR.(string)
+		if !okLR {
+			t.FailNow()
+		}
+		expr, ok := res.(string)
+		if !ok {
+			t.FailNow()
+		}
+		if expr != exprLR {
+			t.Fatalf(
+				"for input %q\ngot result: %q,\nbut expect: %q",
+				data, exprLR, expr)
+		}
+	})
+}
```

Comment:
> One more idea to gain some insights into the robustness of this implementation as well as coverage of edge cases would be to have two grammars for parsing the same input (e.g. arithmetic expressions), one with left recursion and one without and then use fuzzing to compare the results of the two.

I added fuzzing. But I have no idea how to add this to CI. I need help =)

Reply:
Thanks for adding fuzzing. For now, I am ok with not running the fuzzing in CI.